cjdns-admin/src/conn.rs

```diff
@@ -0,0 +1,192 @@
+        }
+
+        Ok(res)
+    }
+
+    /// Call remote function.
+    pub async fn call_func<A: msgs::Args, P: msgs::Payload>(&mut self, remote_fn_name: &str, args: A, disable_auth: bool) -> Result<P, Error> {
+        //dbg!(remote_fn_name);
+
+        if disable_auth || self.password.is_empty() {
+            self.call_func_no_auth(remote_fn_name, args).await
+        } else {
+            self.call_func_auth(remote_fn_name, args).await
+        }
+    }
+
+    async fn call_func_no_auth<A: msgs::Args, P: msgs::Payload>(&mut self, remote_fn_name: &str, args: A) -> Result<P, Error> {
+        let msg = msgs::Query {
+            txid: self.counter.next().to_string(),
+            q: remote_fn_name.to_string(),
+            args,
+        };
+
+        let resp: msgs::GenericResponse<P> = self.send_msg(&msg).await?;
+        check_txid(&msg.txid, &resp.txid)?;
+        check_remote_error(&resp.error)?;
+
+        Ok(resp.payload)
+    }
+
+    async fn call_func_auth<A: msgs::Args, P: msgs::Payload>(&mut self, remote_fn_name: &str, args: A) -> Result<P, Error> {
+        // Ask cjdns for a cookie first
+        let new_cookie = {
+            let resp: msgs::CookieResponsePayload = self.call_func_no_auth("cookie", ()).await?;
+            resp.cookie
+        };
+
+        // Hash password with salt
+        let passwd_hash = {
+            let cookie_passwd = self.password.clone() + &new_cookie;
+            let digest = hash(cookie_passwd.as_bytes());
+            hex::encode(digest)
+        };
+
+        // Prepare message with initial hash
+        let mut msg = msgs::AuthQuery {
+            txid: self.counter.next().to_string(),
+            q: "auth".to_string(),
+            aq: remote_fn_name.to_string(),
+            args,
+            cookie: new_cookie,
+            hash: passwd_hash,
+        };
+
+        // Update message's hash
+        let msg_hash = {
+            let msg_bytes = msg.to_bencode()?;
+            let digest = hash(&msg_bytes);
+            hex::encode(digest)
+        };
+        msg.hash = msg_hash;
+
+        // Send/receive
+        let resp: msgs::GenericResponse<P> = self.send_msg(&msg).await?;
+        check_txid(&msg.txid, &resp.txid)?;
+        check_remote_error(&resp.error)?;
+
+        Ok(resp.payload)
+    }
+
+    async fn send_msg<RQ, RS>(&mut self, req: &RQ) -> Result<RS, Error>
+        where RQ: msgs::Request,
+              RS: msgs::Response
+    {
+        // Send encoded request
+        let msg = req.to_bencode()?;
+        //dbg!(String::from_utf8_lossy(&msg));
+        self.socket.send(&msg).await.map_err(|e| Error::NetworkOperation(e))?;
+
+        // Limit receive packet lenght to typical Ethernet MTU for now; need to check actual max packet length on CJDNS Node side though.
+        let mut buf = [0; 1500];
+
+        // Reseive encoded response synchronously
+        let received = self.socket.recv(&mut buf).await.map_err(|e| Error::NetworkOperation(e))?;
```

Comment:
What if the response doesn't fit in the buf?
We should be flexible. But at the same time there should be a hard limit to prevent DoS.

Reply:
Maximum practical length for UDP packed is slightly less that 1500 bytes (limited to ethernet/wifi frame size), so cjdns router also limits its messages (that's why they have 'page' parameter instead of just single list of functions etc). So this buffer seem to be enough.